### ISSUE TITLE
remove annoying "double callback" warning

### ISSUE THREAD

**Issue by stephenmathieson** [visionmedia/superagent#313]
i'd rather have an error thrown than `console.warn('double callback!')`

**Comment by kfng7** [comment]
I also got the same issue when the remote server is down. I expect that error message should be passed via callback instead.

**Comment by lukebayes** [comment]
Also experiencing this issue using gulp watch with mocha.

```javascript

var request = require('supertest'),
      worker = require('../src/worker');

describe('POST /message', function() {
  var app;

  beforeEach(function() {
    app = worker();   
  });

  it('handles valid request', function(done) {
    request(app)
      .post('/message')
      .send({version: 10, docId: 'blah'})
     .expect(200)
     .end(done);
  });
});
```

This test passes the first time every time. Then after about 3 file saves it will fail for the next few file saves, and eventually pass again intermittently going back and forth with the double callback error.

```
TypeError: Cannot read property 'length' of undefined
    at Test.Request.callback ([project]/node_modules/supertest/node_modules/superagent/lib/node/index.js:746:14)
    at ClientRequest.<anonymous> ([project]/node_modules/supertest/node_modules/superagent/lib/node/index.js:711:10)
    at ClientRequest.EventEmitter.emit (events.js:95:17)
    at Socket.socketOnEnd [as onend] (http.js:1568:9)
    at Socket.g (events.js:180:16)
    at Socket.EventEmitter.emit (events.js:117:20)
    at _stream_readable.js:919:16
    at process._tickDomainCallback (node.js:463:13)
```

**Comment by lukebayes** [comment]
I may have found the root cause and a workaround.

It looks like providing a callback handler to `expect` _and_ adding an `end` handler may intermittently trigger this multiple callback warning.

I have changed our test code to always pass a callback to `expect`, never call `end()` and the provided handler calls `done()` directly.

Haven't seen a recurrence for an hour or so.

**Comment by JohnTigue** [comment]
I found my way here because of "double callback!" messages while running Mocha tests with Nock. I am leaving this message mostly to cross-link to more info at #417.

And in Nock's issues: https://github.com/pgte/nock/issues/211

One small trick for checking to see if the SuperAgent-to-Nock interaction is your problem is to call nock.enableNetConnect() right before firing off a request with SuperAgent. Toggle that on and off with comments while running Mocha and you might just have found where your prey is hiding.

**Comment by SGrondin** [comment]
I have found a workaround for my particular case.

In a mocha test file, I was doing:
```javascript
request.post("http://127.0.0.1/")
.send(new Buffer(1024*1025))
.end(function(err, result) {
	// Some asserts go here
	done()
})
```

I changed it to
```javascript
var buf = new Buffer(1025*1024).toString("utf8")
request.post("http://127.0.0.1/")
.send(buf)
.end(function(err, result) {
	// Some asserts go here
	done()
})
```
and the problem disappeared

**Comment by asperling** [comment]
Same here but with a slightly different environment. I use [supertest-as-promised](https://github.com/WhoopInc/supertest-as-promised) and therefore my tests are promise based.

But my solution might be of interest anyway, since I figured that the *double callback* issue disappeared when I re-instantiated the agent in the `beforeEach()` hook.

```
var expect = require('chai').expect,
    q = require('q'),
    sap = require('supertest-as-promised'),
    app = require('../../server')
    agent,
    request;

// ...

describe('Some Service Tests', function() {
  beforeEach(function() {
    // reinitialize agent to prevent "double callback"s
    request = sap(q.Promise);
    agent = request.agent(app);
    // ...
  }
  // Tests go here... and so on
}
```

Now... I never saw the *double callback* error since.

**Comment by JMacLulich** [comment]
I'm seeing this problem as well, it happens only sporadically but my "end" handler is being called twice on some fairly simple code.

```javascript
request
      .post(url)
      .send(postBody)
      .type('application/json')
      .timeout(TIMEOUT)
      .end(function(err, res) {  ... some stuff gets executed twice for some weird reason... };
```

I'll try the suggestion about using `expect` instead of `end` but not sure why that should matter.

**Comment by ORESoftware** [comment]
I was getting the double callback warning for this

```javascript
     it('Should return 200 when /log_stdout is called.', function (done) {
            request($url)
                .get('/log_stdout')
                .expect(200)
                .end(function (err, res) {
                    if (err) {
                        return done(err);
                    }
                    res.statusCode.should.equal(200);
                    done();
                });
        });
```

and I changed it to:

```javascript
     it('Should return 200 when /log_stdout is called.', function (done) {
            request($url)
                .get('/log_stdout')
                .expect(200)
                .end(function (err, res) {
                    if (err) {
                        return done(err);
                    }
                   else{
                     res.statusCode.should.equal(200);
                    done();
                }
                });
        });
```

IMO using the else is better practice anyway

**Comment by matthiasg** [comment]
@ORESoftware i dont quite follow your code there. Your change should not have corrected this issue as i understand it. 

@JMacLulich writes that `end` is called at least twice, that means in your example you either call the if(err) branch once or twice or the else branch once or twice. In all cases you would call `done` exactly as often as `end` was called. 

So why does that fix anything ?

**Comment by rcrichton** [comment]
I'm also seeing the 'double callback!' message and my `end()` callback doesn't seem to be called (the test times out in mocha). For me it was introduced with supertest 1.1.0 and the same code works fine on 1.0.1. I'm also using mocha for testing and this occurs in a few of our test.

Here is a snippet of code. Is there anything we are doing wrong to cause this?

```coffeescript
describe "with no credentials", ->
  it "should `throw` 401", (done) ->
    server.start httpPort: 5001, ->
      request("http://localhost:5001")
        .get("/test/mock")
        .expect(401)
        .end (err, res) ->
          if err
            done err
          else
            done()
```

**Comment by vizath** [comment]
I'm having this issue also, when making requests with small timeouts (timeouts that are near the response time).
I can reproduce it by executing this snippet.

```
var request = require('superagent');
request.get('http://google.com').timeout(140).end(function(err, res) { console.log(err); });
request.get('http://google.com').timeout(140).end(function(err, res) { console.log(err); });
request.get('http://google.com').timeout(140).end(function(err, res) { console.log(err); });
request.get('http://google.com').timeout(140).end(function(err, res) { console.log(err); });
request.get('http://google.com').timeout(140).end(function(err, res) { console.log(err); });
request.get('http://google.com').timeout(140).end(function(err, res) { console.log(err); });
request.get('http://google.com').timeout(140).end(function(err, res) { console.log(err); });
```
You can make the request only once, but you have more chances to see it if you execute it often.

```
$ node index.js
null
null
null
null
{ [Error: timeout of 140ms exceeded] timeout: 140, code: 'ECONNABORTED', response: undefined }
{ [Error: timeout of 140ms exceeded] timeout: 140, code: 'ECONNABORTED', response: undefined }
{ [Error: timeout of 140ms exceeded] timeout: 140, code: 'ECONNABORTED', response: undefined }
double callback!
double callback!
double callback!
```

**Comment by alexaivars** [comment]
Why is this issue still closed ? Im still having problems with double callback, the following simple situation causes a double callback. It might be related to JSON.parse throwing an error further down the chain.
 
```
		require('nock')('http://foo.io')
			.get('/invalid')
			.reply(200, 'not valid json', {'Content-Type': 'application/json'})
	
		require('superagent')
			.get('http://foo.io/invalid')
			.end(function (err, res) {	
			
			});
```

**Comment by psypersky** [comment]
Same problem here

**Comment by tribou** [comment]
I was seeing the 'double callback!' message along with a `fn is not a function` error when mocking superagent with sinon.  After visionmedia/superagent@454eeb2, the error and double callback message are gone.  Did that commit resolve the issue for anyone else?

**Comment by matthiasg** [comment]
I had this issue again today after doing some extensive unrelated updates in my codebase. In my case the culprit turned out to be the 'faux promise support'.

basically i did something like this (abbreviated): 

```
function send(something,callback){
  return request.post('http://foo.io/invalid').send(something).end(callback);
 }

function main(){
   send(something, function(){ console.log(done) }).then(function(){})
}
```

the `then` was added because i used promises in my main code base and the call to `send` was inside a promise then handler.

```
Promise.all([...]).then( ()=>send(somedata, function(){...}) )
```

this happens in ES6 with the above syntax or with e.g coffeescript or with ES5 when its `function(){return send(...)}` 

the return value of my send function is interpreted as a promise thus the `faux` promise support of `Request` is called causing `end` to be called twice.

in the 2nd call `this._headerSent` is already set to true which means the data is not serialized in the second call causing the  `req.end(data);` to complain about having the wrong format when the send call is given an object.

TL;DR: had the request hooked up with 'end callback' AND '.then callback' causing double request.

**Comment by ybian** [comment]
I had the same issue. The first case of the following works well but the second prints "double callback" message and the done() callback was not called so timeout error occurred.  

It seems that supertest inside a promise then block can easily reproduce the problem.

```coffeescript
  it '# success', (done) ->
    supertest(dbServerUrl)
      .get('/_all_dbs')
      .expect(200)
      .end (err) ->
        if err
          done.fail(err)
        else
          done()

  it '# failed', (done) ->
    Q('5')
    .then ->
      supertest(dbServerUrl)
        .get('/_all_dbs')
        .expect(200)
        .end (err) ->
          if err
            done.fail(err)
          else
            done()
```

**Comment by gstamp** [comment]
@ybian I also get the same issue when I use promises although in my case I was using bluebird.

**Comment by pornel** [comment]
Superagent 2.0 fixes handling of `.then()`

**Comment by vizath** [comment]
As of 1.8.3 and 2.0.0-alpha1, my [test case higher](https://github.com/visionmedia/superagent/issues/313#issuecomment-141326537) still outputs `double callback!`.

**Comment by hustcer** [comment]
I got the `double callback!` warning too.

**Comment by cdutson** [comment]
I'm on 2.0.0-alpha3, and I'm getting double callback

example code:

```
export const performRequest = function performRequest(url, method='GET') {
 return request(method, url);
};

export const getResource = function getResource(resource, id = null) {
    return performRequest(`${baseUrl}/${resource}/${id}`);
};

export const _doFetch = function _doFetch(id, resolve, reject) {
  // Right here it's double callbacking the .end() causes it.
  // I can even run getResource('sample', id).end() to generate the warning, no resolve/reject needed
  return getResource('sample', id).end((err, res) => { 
    if (res && res.ok) {
     resolve(res.body);
    }
    else {
      reject(err.response);
    }
  });
};
```

**Comment by pornel** [comment]
I can't reproduce this issue with any of the code posted in the thread.

I suspect the double callback issue depends on specific HTTP headers and possibly timing of server response, so if you're reporting the issue, please include real, working public URLs that I can investigate.

**Comment by cdutson** [comment]
@pornel further information that may prove helpful, I only see these warnings when running the code through mocha. so some additional code:

```
/* import nock, expect, etc */

describe('api code', () => {
   it('Should return values for api call', () => {
    nock('http://33.0.0.16:8000')
      .get('/foo')
      .reply(200, [
        {
          foo: 'bar'
        }
      ]);

    const expectedResults = [
        {
            foo: 'bar'
        }
    ]

    return _doFetch().then((results) => {
      expect(results).toEqual(expectedResults);
    },
    () => { /* swallow error */ });
  });
}
```

running the code posted above within the test suite causes ... 3 warnings to show up.
```
cli> node_modules/mocha/bin/_mocha --compilers js:babel-register --recursive --grep "api code"

double callback!
double callback!
double callback!
    ✓ Should return values for api call
```

**Comment by romaleev** [comment]
Same issue with 'double callback!' with version 1.8.4.
I use PM2 with node.js.
When superagent's timeout exceeded I log the error and run process.exit after which PM2 successfully restarts the node.
However sometimes superagent sends 'double callback!' warning right after I call process.exit which makes my node hang up forever and not being restarted by PM2.
After some investigation I figured out that request callback executes twice:
1) superagent.callback with error -> which resolves it as expected
2) calling process.exit (but it is still running afterwards)
3) then callback executes the second time !!! with a result and no error (result is with content as if request succeeded)
4) superagent returns 'double callback!' warning within neither callback fn execution nor error.
5) node.js is still running

As a conclusion for some reason superagent executes callback twice with no resolving the second one which leads the process to hang up.

**Comment by pornel** [comment]
We know 1.8 is buggy — we've already fixed those bugs in later versions. Especially 2.2 and later removed many double callback problems. Please upgrade.

**Comment by ajbogh** [comment]
This error happens when you create a function that accepts multiple parameters and use that function within an `it`. In my case I created an "ensureLoggedIn" promise that I can run for all of my tests. The error has nothing to do with the code inside the "ensureLoggedIn" function, it only errors when the function definition has more than one parameter.

```
describe('My Test', function(){
    it('should perform some task after logging in', function(done){
      //will use default args in function
      auth.ensureLoggedIn().then(function(){
        //perform the task
      });
    });
});
```

*auth.js*
```
module.exports = {
  // Specifying any more than one parameter in the function definition will create the error.
  // The alternative is to use arguments[0] and arguments[1] instead of systemId and username, 
  // then leave the parens in the function definition empty. 
  ensureLoggedIn: function(systemId, username){
    systemId = systemId || 100;
    username = username || "myuser";

    if(!global.securityToken){
      var agent = request.agent(config.myService)

      return agent
        .get(config.routes.userToken+'?systemid='+systemId+'&username='+username)
        .end(function(err, res){
          global.userId = res.body.UserId
          global.securityToken = res.body.SecureToken;
        });
    }else{
      return Promise.resolve();
    }
  }
}
```

**Comment by djabraham** [comment]
I was getting double callback because I had the port forwarded via nat using virtual box and (boot2docker). I was running a containerized app locally and forgot the same port was used for both. 

Funny thing is the server starts and I am able to browse to it with no warnings. But when I run supertest, I encounter the double callback error. Once I changed the port, it worked fine. 

I don't consider this a problem, just posting it here for future reference. This was occurring under the latest versions (2.0.0 SA & 2.3.0 ST).

**Comment by pornel** [comment]
Superagent 2.0.0 has this bug. The fix is in Superagent 2.2 or later.

Supertest changes quite a lot about behavior of superagent, so if you have a bug that happens when running supertest, please report it to the supertest project.

**Comment by vizath** [comment]
I can still reproduce https://github.com/visionmedia/superagent/issues/313#issuecomment-141326537 in 3.3 using `.timeout(180)`. 
The warning is now `superagent: double callback bug`.

Note that `.timeout({ response:60 })` (introduced in 3.2) works fine with no warnings, but it is not the behavior that I'm looking for (I don't want to wait more than 180ms total including downloading the body).

Could that mean that the warning is shown when there is an error between the moment the connection is established and the body is completely received ?

**Comment by pornel** [comment]
Sorry, I can't reproduce this (superagent 3.3, Node 7.2 osx). Timeout is in our [test](https://github.com/visionmedia/superagent/blob/master/test/timeout.js) suite.

**Comment by pornel** [comment]
If you run into this problem, please change:

```js
  if (this.called) return console.warn('superagent: double callback bug');
  this.called = true;
```

to

```  
  if (this.called) return console.trace('superagent: double callback bug', this.called);
  this.called = Error("previous trace");
```

and report the stack traces.

**Comment by vizath** [comment]
```
$ node superagent.js
null
null
{ [Error: Timeout of 450ms exceeded] timeout: 450, code: 'ECONNABORTED', response: undefined }
{ [Error: Timeout of 450ms exceeded] timeout: 450, code: 'ECONNABORTED', response: undefined }
{ [Error: Timeout of 450ms exceeded] timeout: 450, code: 'ECONNABORTED', response: undefined }
{ [Error: Timeout of 450ms exceeded] timeout: 450, code: 'ECONNABORTED', response: undefined }
{ [Error: Timeout of 450ms exceeded] timeout: 450, code: 'ECONNABORTED', response: undefined }

Trace: superagent: double callback bug [Error: previous trace]
    at Request.callback ([...]\node_modules\superagent\lib\node\index.js:637:35)
    at Stream.<anonymous> ([...]\node_modules\superagent\lib\node\index.js:849:18)
    at emitNone (events.js:67:13)
    at Stream.emit (events.js:166:7)
    at Unzip.<anonymous> ([...]\node_modules\superagent\lib\node\unzip.js:63:12)
    at emitNone (events.js:72:20)
    at Unzip.emit (events.js:166:7)
    at endReadableNT (_stream_readable.js:905:12)
    at nextTickCallbackWith2Args (node.js:441:9)
    at process._tickCallback (node.js:355:17)
```

You need to play with the timeout values until you get the sweet spot where half of them are timeouts and half of them are success.
Also, having the timeout error doesn't always trigger the trace.

Using a mobile network makes it very hard to reproduce.

**Comment by vizath** [comment]
I cannot reproduce my use case on 3.3.1 :tada:
